### PR TITLE
feat(java_indexer): implement ref/writes edges

### PR DESCRIPTION
**Pull request by wcalandro**
This PR adds support for `ref/writes` edges to the Java indexer. The edges are emitted for variables and class members. See the verifier test for a full representation of when `ref`, `ref/writes`, or both edges are emitted.

### CODE REVIEW

File: kythe/javatests/com/google/devtools/kythe/analyzers/java/testdata/pkg/Writes.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    //
+
+    //- @sub ref MemberSub
+    //- !{ @sub ref/writes MemberSub }
+    //- @#0a ref/writes MemberA
+    //- @#1a ref MemberA
+    sc.sub.a = sc.a;
+    //- @a ref MemberA
+    //- !{ @a ref/writes MemberA }
+    b = sc.sub.a;
```
Comment (jaysachs) [inline kythe/javatests/com/google/devtools/kythe/analyzers/java/testdata/pkg/Writes.java:96]: What about a test of
```
  x = y = z++;
```
(if that's supported)

File: kythe/javatests/com/google/devtools/kythe/analyzers/java/testdata/pkg/Writes.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+  class Subclass {
+    //- @a defines/binding MemberA
+    public int a = 1;
+
+    //- @sub defines/binding MemberSub
+    public Subclass sub = new Subclass();
+
+    public Subclass() {}
+  }
```
Comment (jaysachs) [inline kythe/javatests/com/google/devtools/kythe/analyzers/java/testdata/pkg/Writes.java:107]: Add a method and a test that uses `this`, e.g.
```
  //- @x defines/bindings ArgX
  private void inc(int x) {
     //- @a ref/writes MemberA
     //- @x ref ArgX
     this.a += x;
  }
```

File: kythe/java/com/google/devtools/kythe/analyzers/java/KytheTreeScanner.java [hunk 3]
```diff
@@ -844,11 +853,24 @@ public JavaNode visitTypeApply(JCTypeApply tApply, TreeContext owner) {
...
+      if (imprt != null) {
+        return emitNameUsage(ctx, sym, name, EdgeKind.REF_IMPORTS);
+      }
+
+      RefType refType = getRefType(owner.getTree(), sym, field.pos);
+      // RefType can only be READ, WRITE, OR READ_WRITE so node will always have a value set by
+      // at least one call to emitNameUsage. We just have to initialize to null to keep the compiler
+      // happy.
+      JavaNode node = null;
+      if (refType == RefType.READ || refType == RefType.READ_WRITE) {
```
Comment (jaysachs) [inline kythe/java/com/google/devtools/kythe/analyzers/java/KytheTreeScanner.java:867]: Make this a switch statement? It's odd that both branches match `READ_WRITE`.

File: kythe/java/com/google/devtools/kythe/analyzers/java/KytheTreeScanner.java [hunk 4]
```diff
@@ -290,12 +290,21 @@ public JavaNode visitIdent(JCIdent ident, TreeContext owner) {
...
-    EdgeKind edgeKind = EdgeKind.REF;
+
+    JavaNode node = null;
     if (ident.sym instanceof ClassSymbol && ident == owner.getNewClassIdentifier()) {
       // Use ref/id edges for the primary identifier to disambiguate from the constructor.
-      edgeKind = EdgeKind.REF_ID;
+      node = emitSymUsage(ctx, ident.sym, EdgeKind.REF_ID);
+    } else {
+      RefType refType = getRefType(owner.getTree(), ident.sym, ident.pos);
+      if (refType == RefType.READ || refType == RefType.READ_WRITE) {
```
Comment (jaysachs) [inline kythe/java/com/google/devtools/kythe/analyzers/java/KytheTreeScanner.java:300]: So if it's a READ_WRITE, we emit both the ref and the ref/write? Is there a test for this?

Comment (wcalandro) [reply kythe/java/com/google/devtools/kythe/analyzers/java/KytheTreeScanner.java:300]: Yes, the follow tests this:
```
//- @a ref A
//- @a ref/writes A
a += 1;
```

File: kythe/java/com/google/devtools/kythe/analyzers/java/KytheTreeScanner.java (hunk 3 above)
Comment (wcalandro) [reply kythe/java/com/google/devtools/kythe/analyzers/java/KytheTreeScanner.java:867]: I could potentially do a switch. This is just based on the logic from Cody's implementation for Go.

File: kythe/javatests/com/google/devtools/kythe/analyzers/java/testdata/pkg/Writes.java (hunk 2 above)
Comment (wcalandro) [reply kythe/javatests/com/google/devtools/kythe/analyzers/java/testdata/pkg/Writes.java:115]: Added test

File: kythe/javatests/com/google/devtools/kythe/analyzers/java/testdata/pkg/Writes.java (hunk 1 above)
Comment (wcalandro) [reply kythe/javatests/com/google/devtools/kythe/analyzers/java/testdata/pkg/Writes.java:125]: Chained assignments and variable writes on the right-hand side now work.

**Review by schroederc — APPROVED** [review]

**Pull request merged**